### PR TITLE
Add support for redirecting captured output as it's being written

### PR DESCRIPTION
**Pull request by msanders**
This allows printing or redirecting output as it's being captured, analogous to the `tee` command. For example:

```sh
NSUnbufferedIO=YES softwareupdate --list | tee /dev/tty | grep -Fq $needle
```

Can be written as:

```swift
let result = try Task.capture(
    "softwareupdate",
    arguments: ["--list"],
    outputStream: Term.stdout,
    env: [
        "NSUnbufferedIO": "YES",
    ]
)

if result.output.contains(needle) { /* ... */ }
```

Currently to do this without a fork requires re-implementing `Task.capture`, `CaptureStream`, `CaptureResult`, and `CaptureError`.

I've also added support for passing in `forwardInterrupt` and `env` properties as arguments, can open a separate PR for that if necessary.

For `CaptureStream`, I used a block instead of another `WritableStream` parameter as that was more consistent with the surrounding code and similar to the [`readabilityHandler`](https://developer.apple.com/documentation/foundation/filehandle/1412413-readabilityhandler) block on `FileHandle`.

### CODE REVIEW

**Comment by jakeheis** [comment]
I'm inclined to think that this might be better implemented as more general `SplitStream` that would also allow output to be written to multiple files at once. Something like:

```swift
public class SplitStream: WritableStream {
    
    public let writeHandle: FileHandle
    public let processObject: Any
    public var encoding: String.Encoding = .utf8
    
    private let queue = DispatchQueue(label: "com.jakeheis.SwiftCLI.SplitStream")
    private let semaphore = DispatchSemaphore(value: 0)
    
    public init(streams: [WritableStream]) {
        let pipe = Pipe()
        self.processObject = pipe
        self.writeHandle = pipe.fileHandleForWriting
        
        let readStream = ReadStream.for(fileHandle: pipe.fileHandleForReading)
        queue.async { [weak self] in
            while let data = readStream.readData() {
                streams.forEach { $0.writeData(data) }
            }
            self?.semaphore.signal()
        }
    }
    
    public convenience init(_ streams: WritableStream...) {
        self.init(streams: streams)
    }
    
}

// In use:

let capture = CaptureStream()
let result = Task(
    executable: "/bin/ls",
    arguments: ["."],
    stdout: SplitStream(Term.stdout, capture)
).runSync()
```

Would this accomplish what you're talking about? I think `capture` with the additional arguments you've added could be implemented with this sort of split stream

**Comment by msanders** [comment]
Thanks! That does make this simpler to implement, although unfortunately still requires duplicating an implementation of `CaptureResult` and `CaptureError` to take care of error handling. Would it be possible to make those initializers public instead, or is there something that would fit better with the existing API?

**Comment by jakeheis** [comment]
I think making the initializers public is probably the best call:

```swift
public struct CaptureResult {
     public init(stdout: CaptureStream, stderr: CaptureStream) { ... }
}
public struct CaptureError: ProcessError {
     public init(exitStatus: Int32, captured: CaptureResult) { ... }
}

// In use

let stdoutCap = CaptureStream()
let stderrCap = CaptureStream()
let result = Task(
    executable: "/bin/ls",
    arguments: ["."],
    stdout: SplitStream(Term.stdout, stdoutCap),
    stderr: stderrCap
).runSync()
// Build CaptureResult and CaptureError
```

Would something like that work for you?

**Comment by msanders** [comment]
Thanks, that does solve my use-case. Updated PR with proposed changes.

File: Sources/SwiftCLI/Stream.swift [hunk 1]
```diff
@@ -494,6 +494,40 @@ public class CaptureStream: ProcessingStream {
     
 }
 
+public class SplitStream: WritableStream {
```
Comment (jakeheis) [inline Sources/SwiftCLI/Stream.swift:497]: Needs to be `public class SplitStream: ProcessingStream {`

File: Sources/SwiftCLI/Stream.swift [hunk 2]
```diff
@@ -494,6 +494,40 @@ public class CaptureStream: ProcessingStream {
...
+                $0.closeWrite()
+            }
+            self?.semaphore.signal()
+        }
+    }
+
+    public convenience init(_ streams: WritableStream...) {
+        self.init(streams: streams)
+    }
+
```
Comment (jakeheis) [inline Sources/SwiftCLI/Stream.swift:528]: Needs to have:
```swift
 public func waitToFinishProcessing() {
        streams.forEach { ($0 as? ProcessingStream)?.waitToFinishProcessing() }
        semaphore.wait()
    }
```

**Comment by jakeheis** [comment]
Left a few comments about conforming `SplitStream` to `ProcessingStream`, otherwise looks good!

**Comment by msanders** [comment]
Thanks, done.

**Comment by jakeheis** [comment]
Looks great, thank you for the contribution!

**Comment by jakeheis** [comment]
`6.0.2`

**Pull request merged**